incremental.go

```diff
@@ -0,0 +1,267 @@
+package feedx
+
+import (
+	"context"
+	"errors"
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	"github.com/bsm/bfs"
+)
+
+// IncrementalProduceFunc is a callback which is run by the producer on every iteration.
+type IncrementalProduceFunc func(w *Writer, lastModified time.Time) error
+
+// LastModifiedFunc will be called before each push attempt
+// to determine the last updated record time.
+type LastModifiedFunc func(context.Context) (time.Time, error)
+
+// Manifest describes the current feed status.
+// the current manifest is consumed before each push and a new manifest written after each push.
+type Manifest struct {
+	// LastModified holds a last-modified time of the records included in Files.
+	LastModified timestamp `json:"lastModified"`
+	// Generation is a incrementing counter for use in file compaction.
+	Generation int `json:"generation"`
+	// Files holds a set of incremental data files
+	Files []string `json:"files"`
+}
+
+// IncrementalProducerOptions configure the producer instance.
+type IncrementalProducerOptions struct {
+	WriterOptions
+
+	// The interval used by producer to initiate a cycle.
+	// Default: 1m
+	Interval time.Duration
+
+	// AfterPush callbacks are triggered after each push cycle, receiving
+	// the push state and error (if occurred).
+	AfterPush func(*IncrementalProducerPush, error)
+}
+
+func (o *IncrementalProducerOptions) norm() {
+	if o.Format == nil {
+		o.Format = ProtobufFormat
+	}
+
+	if o.Compression == nil {
+		o.Compression = GZipCompression
+	}
+
+	if o.Interval <= 0 {
+		o.Interval = time.Minute
+	}
+}
+
+// IncrementalProducerPush contains the state of the last push.
+type IncrementalProducerPush struct {
+	// Producer exposes the current producer state.
+	*IncrementalProducer
+	// Updated indicates is the push resulted in an update.
+	Updated bool
+}
+
+// IncrementalProducer produces a continuous incremental feed.
+type IncrementalProducer struct {
+	bucket    bfs.Bucket
+	ownBucket bool
+	opt       IncrementalProducerOptions
+
+	ctx   context.Context
+	stop  context.CancelFunc
+	modfn LastModifiedFunc
+	pfn   IncrementalProduceFunc
+
+	numWritten, lastPush, lastMod int64
+}
+
+// NewIncrementalProducer inits a new incremental feed producer.
+func NewIncrementalProducer(ctx context.Context, bucketURL string, opt *IncrementalProducerOptions, modfn LastModifiedFunc, pfn IncrementalProduceFunc) (*IncrementalProducer, error) {
+	bucket, err := bfs.Connect(ctx, bucketURL)
+	if err != nil {
+		return nil, err
+	}
+
+	p, err := NewIncrementalProducerForBucket(ctx, bucket, opt, modfn, pfn)
+	if err != nil {
+		_ = bucket.Close()
+		return nil, err
+	}
+	p.ownBucket = true
+
+	return p, nil
+}
+
+// NewIncrmentalProducerForRemote starts a new incremental feed producer for a bucket.
+func NewIncrementalProducerForBucket(ctx context.Context, bucket bfs.Bucket, opt *IncrementalProducerOptions, modfn LastModifiedFunc, pfn IncrementalProduceFunc) (*IncrementalProducer, error) {
```

Comment:
Since a lot of this stuff is VERY similar to `Producer`, maybe we should just add an `Incremental bool` option to `ProducerOptions` and use existing struct?

Reply:
not 100% set on this, just trying to find a neater solution than duplicating a lot of code

Reply:
yup, my plan was to duplicate a lot in this first review and then look to share stuff once I had it working.

Reply:
ok, after thinking this through I think I get how to do it all off normal  Producer struct with incrmental option, have started implementing